### PR TITLE
[Enhancement]WebRTCClient self isolation

### PR DESCRIPTION
**Pull request by ipavlidakis**
### 📝 Summary

This revision attempts to better isolate WebRTClient from external components. This will be helpful in the Reconnection v2 work as it will allow us to handle state internally to the WebRTC client.

### 🧪 Manual Testing Notes

- Join a call and verify everything is alright
- Disconnect from wifi and reconnect. Verify everything is alrigh

### ☑️ Contributor Checklist

- [ ] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required)
- [ ] This change follows zero ⚠️ policy (required)
- [x] This change should receive manual QA
- [ ] Changelog is updated with client-facing changes
- [ ] New code is covered by unit tests
- [ ] Comparison screenshots added for visual changes
- [ ] Affected documentation updated (Docusaurus, tutorial, CMS)

### 🎁 Meme

_Provide a funny gif or image that relates to your work on this pull request. (Optional)_

### CODE REVIEW

File: Sources/StreamVideo/Controllers/CallController.swift [hunk 1]
```diff
@@ -402,190 +405,190 @@ class CallController: @unchecked Sendable {
...
-            }
-            webRTCClient?.onSessionMigrationEvent = { [weak self] in
-                self?.handleSessionMigrationEvent()
-            }
-            webRTCClient?.onSessionMigrationCompleted = { [weak self] in
-                self?.call?.update(reconnectionStatus: .connected)
-            }
-        }
-        
+//        if let migratingFrom {
```
Comment (martinmitrevski) [inline Sources/StreamVideo/Controllers/CallController.swift:408]: is this a WIP?

File: Sources/StreamVideo/Controllers/CallController.swift [hunk 2]
```diff
@@ -679,50 +682,127 @@ class CallController: @unchecked Sendable {
...
                 await call?.state.update(statsReport: stats)
-                try await webRTCClient?.sendStats(report: stats)
+                try await webRTCClient.sendStats(report: stats)
             } catch {
                 log.error(error)
             }
         }
     }
+
+    fileprivate func _joinWebRTCClient(
```
Comment (martinmitrevski) [inline Sources/StreamVideo/Controllers/CallController.swift:698]: I don't think we should use _ for the private methods, what's the reasoning behind it?

File: Sources/StreamVideo/Controllers/CallController.swift [hunk 3]
```diff
@@ -679,50 +682,127 @@ class CallController: @unchecked Sendable {
...
-                sessionID: $6,
-                ownCapabilities: $7,
-                videoConfig: $8,
-                audioSettings: $9,
-                environment: $10
-            )
+extension CallController: WebRTCClientDelegate {
+    func webRTClientMigrating(_ webRTCClient: WebRTCClient) {
+        Task { @MainActor [weak self] in
+            self?.call?.state.reconnectionStatus = .migrating
```
Comment (martinmitrevski) [inline Sources/StreamVideo/Controllers/CallController.swift:765]: why is this one different than the rest?

File: Sources/StreamVideo/WebRTC/WebRTCClient.swift [hunk 4]
```diff
@@ -154,8 +163,9 @@ class WebRTCClient: NSObject, @unchecked Sendable {
     private(set) var signalChannel: WebSocketClient?
     
     private(set) var sessionID: String
-    private var token: String
     
+    private var token: String = ""
```
Comment (martinmitrevski) [inline Sources/StreamVideo/WebRTC/WebRTCClient.swift:167]: This is not ideal, but I guess it's the consequence of the self isolation.

**Review by martinmitrevski — COMMENTED** [review]
looks good 👍  Before testing (and doing another round of review) - it would be great if we get rid of the commented out code - it's a bit confusing now to see whether something is WIP. Also it's not very readable atm.

File: Sources/StreamVideo/Controllers/CallController.swift (hunk 1 above)
Comment (ipavlidakis) [reply Sources/StreamVideo/Controllers/CallController.swift]: I only kept it in to help us identify if I missed something.

File: Sources/StreamVideo/Controllers/CallController.swift (hunk 2 above)
Comment (ipavlidakis) [reply Sources/StreamVideo/Controllers/CallController.swift:539]: Those methods will be removed during the reconnect implementation. They have the underscore to help me (us) remember that they require special handling.

File: Sources/StreamVideo/Controllers/CallController.swift (hunk 3 above)
Comment (ipavlidakis) [reply Sources/StreamVideo/Controllers/CallController.swift:584]: Just copied what we had (and different handling was there already). Unified it for now

**Comment by ipavlidakis** [comment]
Keep in mind that this is an interim step and it won't land on develop but on `feature/reconnect-v2`.

File: Sources/StreamVideo/WebRTC/WebRTCClient.swift (hunk 4 above)
Comment (martinmitrevski) [reply Sources/StreamVideo/WebRTC/WebRTCClient.swift:167]: not sure I understand the last part - how can you attempt to join a call without calling join? If that's somehow possible, we should try not to crash.

File: Sources/StreamVideo/WebRTC/WebRTCClient.swift [hunk 5]
```diff
@@ -1612,19 +1613,180 @@ class WebRTCClient: NSObject, @unchecked Sendable {
...
+            )
+        }
+    }
+
+    func migrate(
+        response: JoinCallResponse
+    ) {
+        migratingToken = response.credentials.token
+        ownCapabilities = response.ownCapabilities
+        // TODO: Shall we also update audioSettings here?
```
Comment (martinmitrevski) [inline Sources/StreamVideo/WebRTC/WebRTCClient.swift:1646]: in theory, they should be the same

**Pull request closed without merging**